scipy/_lib/_array_api.py

```diff
@@ -195,6 +196,109 @@ def _asarray(
+    #     # These libraries only support one type of device
+    #     return xp_out.__array_namespace_info__().default_device()
+    # else:
+    #     return xp_in.device(array)
+
+    # but there are many issues with device support right now. For example:
+    # - NumPy `from_dlpack` only understands device='cpu' (not a device object)
+    # - CuPy `from_dlpack` doesn't work
+    # - the format of the return of `device` seems different in different libraries
+
+    # So in the meantime:
+    if is_numpy(xp_out):
+        return "cpu"
+    elif is_array_api_strict(xp_out):
+        return xp_out.__array_namespace_info__().default_device()
+    else:
+        return None
+
+
+def xp_asarray(
+        array: ArrayLike,
+        dtype: Any = None,
+        order: Literal['K', 'A', 'C', 'F'] | None = None,
+        copy: bool | None = None,
+        *,
+        xp: ModuleType | None = None,
+        device: str | None = None,
+        check_finite: bool = False,
+        subok: bool = False,
+    ) -> Array:
+    """SciPy-specific replacement for `xp.asarray` with `order`, `check_finite`,
+    `subok`, and automatic array type conversion and device transfer.
+
+    Memory layout parameter `order` is not exposed in the Array API standard.
+    `order` is only enforced if the input array implementation
+    is NumPy based, otherwise `order` is just silently ignored.
+
+    `check_finite` is also not a keyword in the array API standard; included
+    here for convenience rather than that having to be a separate function
+    call inside SciPy functions.
+
+    `subok` is included to allow this function to preserve the behaviour of
+    `np.asanyarray` for NumPy based inputs.
+
+    """
+    xp_in = array_namespace(array)
+    if xp is None:
+        xp = xp_in
+
+    if is_numpy(xp_in):
+        # If object is array-like (but not array), make it an ndarray; otherwise, no-op.
+        array = np.asanyarray(array)
+
+    # Start with special cases that shouldn't be needed, but are due to
+    # shortcomings in library implementations of `device` and `from_dlpack`
+    if is_cupy(xp) or is_torch(xp):
+        # Torch from_dlpack doesn't work with any library - not even torch to torch
+        # CuPy `from_dlpack` doesn't accept a device argument, and when calling
+        # without one, it recommends:
+        # Use `cupy.array(numpy.from_dlpack(input))` instead.
+        array = xp.asarray(np_compat.from_dlpack(array).copy(), dtype=dtype, copy=copy)
+        # I don't know if other libraries support `from_dlpack`, either
+        # We might need other special cases like this.
+    elif is_numpy(xp) and is_torch(xp_in):
+        # NumPy.from_dlpack doesn't work with Torch
+        # Otherwise: Tensor.__dlpack__() got an unexpected keyword argument 'dl_device'
+        array = xp.asarray(array, dtype=dtype, copy=copy)
+    elif is_numpy(xp_in) and not array.__class__.__name__.endswith('.ndarray') and subok:
```

Comment:
```suggestion
    elif is_numpy(xp_in) and not array.__class__.__name__.endswith('ndarray') and subok:
```
For PEP8. Hopefully subclass names don't end with `ndarray`?

Or is there a better way to look for strict array subclasses?

Reply:
You can just use `not (type(array) is np.ndarray)`, I'm pretty sure. `isinstance` accepts subclass relationships, but `type(...) is` does not.